pkgs/development/compilers/djgpp/default.nix

```diff
@@ -0,0 +1,137 @@
+    hash = "sha256-RVfftsFh0yZoCuX6cfAJisSUJaGxG5CgILgxYutwXdo=";
+  };
+
+  djcrossgcc = fetchurl {
+    url = "${DJGPP_DOWNLOAD_BASE}/djgpp/rpms/djcross-gcc-${GCC_VERSION}/djcross-gcc-${GCC_VERSION}.tar.bz2";
+    hash = "sha256-Z5gjk0EcV3m0DOaQED5/RrsT4HYBz1GwEHpHgv6+Ryg=";
+  };
+
+  gcc = fetchurl {
+    url = "${FTPMIRROR_GNU_DOWNLOAD_BASE}/gcc/gcc-${GCC_VERSION}/gcc-${GCC_VERSION}.tar.xz";
+    hash = "sha256-Yv1jSInzHAK2SvLEaPBktHrRynhBHEWr5qxLX43RnHs=";
+  };
+
+  gmp = fetchurl {
+    url = "${FTPMIRROR_GNU_DOWNLOAD_BASE}/gmp/gmp-${GMP_VERSION}.tar.xz";
+    hash = "sha256-/UgpkSzd0S+EGBw0Ucx1K+IkZD6H+sSXtp7d2txJtPI=";
+  };
+
+  mpfr = fetchurl {
+    url = "${FTPMIRROR_GNU_DOWNLOAD_BASE}/mpfr/mpfr-${MPFR_VERSION}.tar.xz";
+    hash = "sha256-DJij8XMv9spOppBVIHnanFl4ctMOluwoQU7iPJVVin8=";
+  };
+
+  mpc = fetchurl {
+    url = "${FTPMIRROR_GNU_DOWNLOAD_BASE}/mpc/mpc-${MPC_VERSION}.tar.gz";
+    hash = "sha256-F1A9LDld/PEGtiLcFCaDwRmUMdCVNnxqrLpu7DA0BFk=";
+  };
+
+  autoconf = fetchurl {
+    url = "${FTP_GNU_DOWNLOAD_BASE}/autoconf/autoconf-${AUTOCONF_VERSION}.tar.xz";
+    hash = "sha256-ZOvOyfisWySHElqGp3YNJZGsnh09vVlIljP53mKldoQ=";
+  };
+
+  automake = fetchurl {
+    url = "${FTP_GNU_DOWNLOAD_BASE}/automake/automake-${AUTOMAKE_VERSION}.tar.xz";
+    hash = "sha256-r2ujkUIiBofFAPebSqLxgdmyTk+NjsSXzqS6JsZL7a8=";
+  };
+
+  # i686 patches from https://github.com/andrewwutw/build-djgpp/issues/45#issuecomment-1484010755
+  patchPhase = ''
+    for f in "build-djgpp.sh" "script/${version}" "setenv/copyfile.sh"; do
+      substituteInPlace "$f" --replace '/usr/bin/env' '${buildPackages.coreutils}/bin/env'
+    done
+  '' + lib.optionalString i686Build ''
+    sed -e 's/i586/i686/g' -e 's:"Building DXE tools.":"Building DXE tools." ; for Z in src/{makefile.def,dxe/makefile.dxe} ; do ( sed "s/i586/i686/g" $Z > $Z-i686 ; mv -v $Z{-i686,} ) ; done :g' script/${version} >script/${buildVersion}
+    chmod -v 755 script/${buildVersion}
+    sed -e 's/i586/i686/g' setenv/setenv >setenv/setenv-i686
+    mv -v setenv/setenv{-i686,}
+  '';
+
+  buildInputs = [ bison curl file flex makeWrapper perl texinfo unzip which ];
+
+  hardeningDisable = [ "format" ];
+
+  buildPhase = ''
+    (mkdir download; cd download
+      ln -s "${autoconf}"   "${baseNameOf autoconf.url}"
+      ln -s "${automake}"   "${baseNameOf automake.url}"
+      ln -s "${binutils}"   "${baseNameOf binutils.url}"
+      ln -s "${djcrossgcc}" "${baseNameOf djcrossgcc.url}"
+      ln -s "${djcrx}"      "${baseNameOf djcrx.url}"
+      ln -s "${djdev}"      "${baseNameOf djdev.url}"
+      ln -s "${djlsr}"      "${baseNameOf djlsr.url}"
+      ln -s "${gcc}"        "${baseNameOf gcc.url}"
+      ln -s "${gmp}"        "${baseNameOf gmp.url}"
+      ln -s "${mpc}"        "${baseNameOf mpc.url}"
+      ln -s "${mpfr}"       "${baseNameOf mpfr.url}"
+    )
+    DJGPP_PREFIX=$out ./build-djgpp.sh ${buildVersion}
+  '';
+
+  postInstall = ''
+    cd $out/*-pc-msdosdjgpp/bin
```

Comment:
I suggest pushd/popd here.

Also, what is this directory? Shouldn't it be automagically set to $out/bin?

Reply:
The script generates some additional tools beyond the usual gcc ones, (ld, strip, etc.) and they end up in a secondary directory. The 'dxe*' ones are important so I lift them out here.

```
❯ find /nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0 | grep bin/ | sort
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/dxe3gen
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/.dxe3gen-wrapped
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/dxe3res
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/dxegen
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/.dxegen-wrapped
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/exe2coff
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-addr2line
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-ar
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-as
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-c++
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-c++filt
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-cpp
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-elfedit
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-g++
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-gcc
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-gcc-12.1.0
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-gcc-ar
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-gcc-nm
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-gcc-ranlib
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-gcov
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-gcov-dump
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-gcov-tool
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-gprof
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-ld
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-ld.bfd
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-lto-dump
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-nm
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-objcopy
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-objdump
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-ranlib
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-readelf
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-size
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-strings
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/i686-pc-msdosdjgpp-strip
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/bin/stubify
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/addr2line
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/ar
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/as
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/c++
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/c++filt
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/cpp
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/dxe3gen
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/dxe3res
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/dxegen
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/exe2coff
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/g++
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/gcc
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/ld
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/ld.bfd
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/nm
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/objcopy
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/objdump
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/ranlib
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/readelf
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/size
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/strings
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/strip
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/stubedit
/nix/store/sj11mv7k3yyjgvvcd8959zl07n6c5jvw-djgpp-12.1.0/i686-pc-msdosdjgpp/bin/stubify
```

Reply:
updated to use clearer logic